src/dbnode/persist/fs/seek.go

```diff
@@ -383,55 +369,106 @@ func (s *seeker) SeekByIndexEntry(entry IndexEntry) (checked.Bytes, error) {
 	return buffer, nil
 }
 
-func (s *seeker) SeekIndexEntry(id ident.ID) (IndexEntry, error) {
-	offset, err := s.indexLookup.getNearestIndexFileOffset(id)
+// SeekIndexEntry performs the following steps:
+//
+//     1. Go to the indexLookup and it will give us an offset that is a good starting
+//        point for scanning the index file.
+//     2. Seek to the position that the indexLookup gave us.
+//     3. Reset a decoder with fileDecoderStream (fd wrapped in a bufio.Reader).
+//     4. Called DecodeIndexEntry in a tight loop (which will advance our position in the
+//        file internally) until we've either found the entry we're looking for or gone so
+//        far we know it does not exist.
+func (s *seeker) SeekIndexEntry(
+	id ident.ID,
+	resources ReusableSeekerResources,
+) (IndexEntry, error) {
+	offset, err := s.indexLookup.getNearestIndexFileOffset(id, resources)
 	// Should never happen, either something is really wrong with the code or
-	// the file on disk was corrupted
+	// the file on disk was corrupted.
+	if err != nil {
+		return IndexEntry{}, err
+	}
+
+	seekedOffset, err := s.indexFd.Seek(offset, os.SEEK_SET)
 	if err != nil {
 		return IndexEntry{}, err
 	}
+	if seekedOffset != offset {
+		return IndexEntry{}, instrument.InvariantErrorf(
+			"tried to seek to offset: %d, but seeked to: %d", seekedOffset, offset)
+	}
 
-	stream := msgpack.NewDecoderStream(s.indexMmap[offset:])
-	s.decoder.Reset(stream)
+	resources.fileDecoderStream.Reset(s.indexFd)
+	resources.xmsgpackDecoder.Reset(resources.fileDecoderStream)
 
 	idBytes := id.Bytes()
-	// Prevent panic's when we're scanning to the end of the buffer
-	for stream.Remaining() != 0 {
-		entry, err := s.decoder.DecodeIndexEntry()
-		// Should never happen, either something is really wrong with the code or
-		// the file on disk was corrupted
+	for {
+		currOffset, err := s.indexFd.Seek(0, os.SEEK_CUR)
```

Comment:
Is this call necessary?  Seems this is not needed because `Seek(0, SEEK_CUR)` should cause no change as `SEEK_CUR` means "seek relative to the current offset",

Reply:
Can't we just keep track of `currOffset` ourselves and avoid this syscall?

Reply:
I have looked at the different interfaces, looks difficult to actually call like `.xmsgpackDecoder.Position()` or whatever which can call the reader underneath.

Maybe just add a comment here that in the future we can remove this Seek (and consequently syscall) call by tracking the position ourselves with some refactoring of interfaces/etc?